src/middleware.page.ts

```diff
@@ -0,0 +1,16 @@
+import { NextRequest, NextResponse } from 'next/server';
+
+// run only on homepage
+export const config = {
+  matcher: '/',
+};
+
+export async function middleware(req: NextRequest) {
+  const { nextUrl: url, geo } = req;
+  console.log(geo);
+  if (!url.searchParams.has('lat') && geo && geo.latitude && geo.longitude) {
```

Comment:
FYI – 0,0 is a valid lat/lng, so it’s probably best to check whether these are `undefined`.

Alternatively if you want absolute type-safety using `zod`, you could go down this route:

```ts
import z from 'zod';

const geoSchema = z.object({
  latitude: z.number();
  longitude: z.number();
})

const geoResult = schema.safeParse(geo);
if(geoResult.success && !url.searchParams.has('center')) {
  const { latitude, longitude } = geoResult.data;
  url.searchParams.set('center', [longitude, latitude].join(','));
}

const centerSchema = z.string()
  .transform((value) => value.split(",").map((l) => parseFloat(l)))
  .pipe(z.tuple([z.number().finite(), z.number().finite()]));
  
const centerResult = centerSchema.parse(router.query.center)
if(centerResult.success) {
  const [lng, lat] = centerResult.data;
  console.log({ lng, lat });
}
```